### PR TITLE
[CSM-296] Cache info about 'used' and 'change' addresses 

### PR DESCRIPTION
**Pull request by Martoon-00**

### CODE REVIEW

File: src/Pos/Wallet/Web/State/Storage.hs [hunk 1]
```diff
@@ -184,6 +205,23 @@ getNextUpdate = preview (wsReadyUpdates . _head)
...
+getCustomAddresses t = HM.keys <$> view (customAddressL t)
+
+getCustomAddress :: CustomAddressType -> CId Addr -> Query (Maybe HeaderHash)
+getCustomAddress t addr = view $ customAddressL t . at addr
+
+addCustomAddress :: CustomAddressType -> (CId Addr, HeaderHash) -> Update Bool
+addCustomAddress t (addr, hh) = fmap isJust $ customAddressL t . at addr <<.= Just hh
+
+removeCustomAddress :: CustomAddressType -> (CId Addr, HeaderHash) -> Update Bool
+removeCustomAddress t (addr, hh) = do
```
Comment (Martoon-00) [inline src/Pos/Wallet/Web/State/Storage.hs]: Try to use `non' (only hh)`

**Comment by akegalj** [comment]
@Martoon-00 @pva701 

Testing:

1) getting wallets when there are no wallets is instant
```
[akegalj@notebook daedalus]$ node
> var api = require('../output/Daedalus.ClientApi')
undefined
> api.getWallets().then(console.log).catch(console.log)
Promise {
  <pending>,
  domain: 
   Domain {
     domain: null,
     _events: { error: [Function: debugDomainError] },
     _eventsCount: 1,
     _maxListeners: undefined,
     members: [] } }
> []
```

2) new wallet takes 5+ sec to response back
```
> api.newWallet('CWTPersonal', 'CWANormal', 1, "science buyer apart miracle medal major kitchen path foil ritual viable pupil", 'pass').then(console.log).catch(console.log)
Promise {
  <pending>,
  domain: 
   Domain {
     domain: null,
     _events: { error: [Function: debugDomainError] },
     _eventsCount: 1,
     _maxListeners: undefined,
     members: [] } }
> { cwPassphraseLU: 1498206806.863057,
  cwMeta: { cwUnit: 1, cwName: 'CWTPersonal', cwAssurance: 'CWANormal' },
  cwId: '1fZk7gQQEh5ZsWcRZfhssrUYmbrsM6xMo1DEU6dgpDtAH9X',
  cwHasPassphrase: true,
  cwAmount: { getCCoin: '0' },
  cwAccountsNumber: 0 }
```

3) calling get wallets after we have created a wallet (see 2) above) takes 5+ sec to return:
```
> api.getWallets().then(console.log).catch(console.log)
Promise {
  <pending>,
  domain: 
   Domain {
     domain: null,
     _events: { error: [Function: debugDomainError] },
     _eventsCount: 1,
     _maxListeners: undefined,
     members: [] } }
> 
> [ { cwPassphraseLU: 1498206806.863057,
    cwMeta: { cwUnit: 1, cwName: 'CWTPersonal', cwAssurance: 'CWANormal' },
    cwId: '1fZk7gQQEh5ZsWcRZfhssrUYmbrsM6xMo1DEU6dgpDtAH9X',
    cwHasPassphrase: true,
    cwAmount: { getCCoin: '0' },
    cwAccountsNumber: 1 } ]
```

3) importing a wallet takes 20+ sec and returns strange error that shouldn't appear to my knowledge:
```
> api.importWallet('/home/akegalj/Downloads/genesis-qanet-2017-05-30/secrets/secret-102.key', 'bla').then(console.log).catch(console.log)
Promise {
  <pending>,
  domain: 
   Domain {
     domain: null,
     _events: { error: [Function: debugDomainError] },
     _eventsCount: 1,
     _maxListeners: undefined,
     members: [] } }
> Error: ServerError: Pos.Wallet.Web.Error.Types.RequestError "Not enough money (need 47682296155 coin(s) more)"
```

I am using keys downloaded from https://input-output-rnd.slack.com/files/george.agapov/F5QLHEXT4/genesis-qanet-2017-05-30-certs-secrets.tgz . Project was compiled with `scripts/build/cardano-sl.sh --prod` and run with `scripts/launch/qa.sh`. Before running I have deleted db's: `rm -rf db-qanet/  wdb-qanet/` .

If you need more details how to use client-api to test see https://cardanodocs.com/technical/wallet-frontend/ or ask @akegalj

**Comment by Martoon-00** [comment]
I'm sorry, looks I forgot to put one of the flags on new evaluation way :pensive: 
Let we fix and test in production now

**Comment by Martoon-00** [comment]
It now works instantly in production, but both flags are evaluated incorrectly for me (it's enough to send money once)

**Comment by Martoon-00** [comment]
Tested in cases of:
* Sending money within same account
* Sending money among wallets forth and back

**Comment by Martoon-00** [comment]
It's inefficient in some sort, I remain https://issues.serokell.io/issue/CSM-324 to do on next week

**Comment by gromakovsky** [comment]
Don't forget to merge `-0.5-rc` to `master.

**Comment by akegalj** [comment]
@Martoon-00 what is this about 
> but both flags are evaluated incorrectly for me
?

**Comment by gromakovsky** [comment]
When will `-0.5-rc` be merged? It currently blocks further merges.

**Comment by Martoon-00** [comment]
@akegalj That's been fixed

@gromakovsky I'm going to do now

**Pull request merged**